core/src/main/java/org/apache/struts2/interceptor/ActionMappingParametersInterceptor.java

```diff
@@ -100,7 +102,12 @@ protected HttpParameters retrieveParameters(ActionContext ac) {
     @Override
     protected void addParametersToContext(ActionContext ac, Map<String, ?> newParams) {
         HttpParameters previousParams = ac.getParameters();
-        HttpParameters.Builder combinedParams = HttpParameters.create().withParent(previousParams).withExtraParams(newParams);
+
+        // TODO can we just change the method signature to Map<String, Parameter>?
```

Comment:
I'm ok with changing the signature here

Reply:
The original code looks like it would survive a _null_ `newParams`, but the updated code could throw a NullPointerException where the old code wouldn't.  Might need a guard for that condition ...

Does newParams _always_ contain double-wrapped `Parameter` instances ?  If not then the new code might fail (I'm not the best with lambda expressions).

If the incoming `Map` always contains double-wrapped `Parameter` values is there a reason why they are double-wrapped ?  If not then maybe the "upstream" double-wrapping should be corrected somewhere further up in the call sequence ...

Reply:
the incoming map is single-wrapped; this is the spot that wrapping the already-wrapped objects